pmlc/rt/level_zero/level_zero_wrappers.cc

```diff
@@ -0,0 +1,250 @@
+                          static_cast<uint32_t>(lws1),
+                          static_cast<uint32_t>(lws2)};
+  void *result =
+      static_cast<LevelZeroInvocation *>(invocation)
+          ->enqueueKernel(static_cast<LevelZeroKernel *>(kernel), gws, lws);
+  IVLOG(2, "  ->" << result);
+  return result;
+}
+/*
+void *levelZeroBarrier(void *invocation, uint32_t count, ...) {
+  std::vector<LevelZeroEvent *> dependencies;
+  va_list args;
+  va_start(args, count);
+  for (unsigned i = 0; i < count; ++i)
+    dependencies.push_back(va_arg(args, LevelZeroEvent *));
+  va_end(args);
+  return static_cast<LevelZeroInvocation *>(invocation)
+      ->enqueueBarrier(dependencies);
+}
+*/
+void level_zero_submit(void *invocation) {
+  IVLOG(2, "level_zero_submit env=" << invocation);
+  static_cast<LevelZeroInvocation *>(invocation)->flush();
+}
+
+void *level_zero_alloc(void *invocation, size_t bytes) {
+  IVLOG(2, "level_zero_alloc env=" << invocation << ", size=" << bytes);
+  void *result =
+      static_cast<LevelZeroInvocation *>(invocation)->allocateMemory(bytes);
+  IVLOG(2, "  ->" << result);
+  return result;
+}
+
+void level_zero_dealloc(void *invocation, void *memory) {
+  IVLOG(2, "level_zero_dealloc env=" << invocation << ", buffer=" << memory);
+  static_cast<LevelZeroInvocation *>(invocation)
+      ->deallocateMemory(static_cast<LevelZeroMemory *>(memory));
+}
+
+void *level_zero_schedule_read(void *host, void *dev, void *invocation,
+                               uint32_t count, ...) {
+  std::vector<LevelZeroEvent *> dependencies;
+  va_list args;
+  va_start(args, count);
+  for (unsigned i = 0; i < count; ++i) {
+    LevelZeroEvent *evt = va_arg(args, LevelZeroEvent *);
+    IVLOG(2, "  event=" << evt);
+    dependencies.push_back(evt);
+  }
+  va_end(args);
+  void *result = static_cast<LevelZeroInvocation *>(invocation)
+                     ->enqueueRead(static_cast<LevelZeroMemory *>(dev), host,
+                                   dependencies);
+  IVLOG(2, "  ->" << result);
+  return result;
+}
+
+void *level_zero_schedule_write(void *host, void *dev, void *invocation,
+                                uint32_t count, ...) {
+  IVLOG(2, "level_zero_schedule_write env=" << invocation << ", host=" << host
+                                            << ", dev=" << dev);
+  std::vector<LevelZeroEvent *> dependencies;
+  va_list args;
+  va_start(args, count);
+  for (unsigned i = 0; i < count; ++i) {
+    LevelZeroEvent *evt = va_arg(args, LevelZeroEvent *);
+    IVLOG(2, "  event=" << evt);
+    dependencies.push_back(evt);
+  }
+  va_end(args);
+  void *result = static_cast<LevelZeroInvocation *>(invocation)
+                     ->enqueueWrite(static_cast<LevelZeroMemory *>(dev), host,
+                                    dependencies);
+  IVLOG(2, "  ->" << result);
+  return result;
+}
+
+void level_zero_wait(uint32_t count, ...) {
+  // No submit created in IR, so call by hand
+  level_zero_submit(instance);
```

Comment:
So we are submit at every comp.Wait Op? Not sure of the overhead, this may be something we should keep in mind when tuning the performance.

Reply:
In L0, each command shall be executed by hand and the command list needs to be reset to be use again,  so do submit before wait to avoid too much submit.  May tune it based on comp dialect.